File: go/lower_test.go
package aontu

import (
	"reflect"
	"strings"
	"testing"
)


func lowerTestCtx(family string, profile map[string]any) *lowerCtx {
	p := map[string]any{"lang": family}
	for k, v := range profile {
		p[k] = v
	}
	lossy := []RenderLoss{}
	return &lowerCtx{profile: p, family: family, unit: "u", lossy: &lossy}
}

func TestLowerWordsAndCases(t *testing.T) {
	for _, c := range []struct {
		name string
		want []string
	}{
		{"HTTPServer2Go", []string{"HTTP", "Server", "2", "Go"}},
		{"utf8_string-value", []string{"utf", "8", "string", "value"}},
		{"naïveName", []string{"naïve", "Name"}},
		{"a b", []string{"a", "b"}},
		{"__", []string{}},
	} {
		if got := lowerSplitWords(c.name); !reflect.DeepEqual(got, c.want) {
			t.Fatalf("%q: want %v, got %v", c.name, c.want, got)
		}
	}
	// A code point at or above U+0080 is never split and never
	// converted: it rides into its word, whatever the style.
	for _, c := range []struct{ name, style, want string }{
		{"crème_brûlée", "pascal", "CrèmeBrûlée"},
		{"über2Go", "snake", "über_2_go"},
		{"a-b c", "kebab", "a-b-c"},
		{"a-b c", "screaming", "A_B_C"},
		{"__", "pascal", "__"},
		{"AsIs_x", "as-is", "AsIs_x"},
	} {
		if got := lowerCaseName(c.name, c.style, nil); got != c.want {
			t.Fatalf("%q %s: want %q, got %q", c.name, c.style, c.want, got)
		}
	}
	if got := lowerCaseName("ledgerId", "camel", []string{"ID"}); "ledgerID" != got {
		t.Fatalf("camel acronym: %q", got)
	}
	if got := lowerCaseName("id", "camel", []string{"ID"}); "id" != got {
		t.Fatalf("camel first word: %q", got)
	}
}

func TestLowerQuote(t *testing.T) {
	profile := map[string]any{"str": map[string]any{
		"quote": "'", "escape": map[string]any{"10": "\\n"},
	}}
	if got := lowerQuote("it's \"x\"\n\x02", profile); "'it\\u0027s \"x\"\\n\\u0002'" != got {
		t.Fatalf("quote: %q", got)
	}
	// No str at all: the double quote, and a bare table.
	if got := lowerQuote("a\"b\\c", map[string]any{}); "\"a\\u0022b\\u005cc\"" != got {
		t.Fatalf("bare: %q", got)
	}
}

func TestLowerLiterals(t *testing.T) {
	goCtx := lowerTestCtx("go", nil)
	tsCtx := lowerTestCtx("typescript", nil)
	for _, c := range []struct {
		v    any
		ctx  *lowerCtx
		want string
	}{
		{1.5, goCtx, "1.5"},
		{int64(3), goCtx, "3"},
		{nil, goCtx, "nil"},
		{nil, tsCtx, "null"},
		{false, tsCtx, "false"},
		{true, tsCtx, "true"},
		{"s", tsCtx, "\"s\""},
	} {
		if got := lowerLiteral(c.v, c.ctx); got != c.want {
			t.Fatalf("%v: want %q, got %q", c.v, c.want, got)
		}
	}
}

func TestLowerLitPrimInGo(t *testing.T) {
	ctx := lowerTestCtx("go", nil)
	for _, c := range []struct {
		of   []any
		want string
	}{
		{[]any{1.5, int64(2)}, "float"},
		{[]any{int64(1), 2.0}, "int"},
		{[]any{true}, "bool"},
		{[]any{nil}, "null"},
		{[]any{"a", int64(1)}, "any"},
	} {
		got := lowerTypeExpr(map[string]any{"k": "lit", "n": c.of}, ctx, "$").text
		if got != c.want {
			t.Fatalf("%v: want %q, got %q", c.of, c.want, got)
		}
	}
	if 5 != len(*ctx.lossy) {
		t.Fatalf("losses: %d", len(*ctx.lossy))
	}
}

func TestLowerFallbacksWithoutTypeForms(t *testing.T) {
	// A supplied profile may name a lowering and no forms: the
	// primitive is its own name, and a form is open-less and close-less.
	ctx := lowerTestCtx("typescript", nil)
	prim := func(p string) map[string]any { return map[string]any{"k": "prim", "prim": p} }
	for _, c := range []struct {
		t    map[string]any
		want string
	}{
		{prim("int"), "int"},
		{map[string]any{"k": "list", "n": prim("int")}, "int"},
		{map[string]any{"k": "map", "key": prim("string"), "n": prim("int")}, "string, int"},
		{map[string]any{"k": "union", "n": []any{prim("a"), prim("b")}}, "a | b"},
	} {
		if got := lowerTypeExpr(c.t, ctx, "$").text; got != c.want {
			t.Fatalf("%v: want %q, got %q", c.t, c.want, got)
		}
	}
}

func TestLowerParenRule(t *testing.T) {
	// The vocabulary keeps a container to leaves, so a list of a
	// nullable reaches the fold only through RenderValue -- where the
	// TypeScript forms say (string | null)[] and not string | null[].
	profile := bundledProfile("typescript")
	report := RenderValue(map[string]any{
		"aontu": map[string]any{"Code": map[string]any{
			"units": []any{map[string]any{
				"path": "a.ts", "lang": "typescript",
				"decls": []any{map[string]any{
					"k": "record", "name": "T", "open": false, "check": []any{},
					"fields": []any{map[string]any{
						"name": "a", "optional": false,
						"type": map[string]any{"k": "list", "n": map[string]any{
							"k": "opt", "n": map[string]any{"k": "prim", "prim": "string"},
						}},
					}},
				}},
			}},
		}},
	}, &RenderOptions{Profiles: []map[string]any{profile}})
	if "ok" != report.Verdict {
		t.Fatalf("verdict: %+v", report)
	}
	if want := "export interface T {\n  a: (string | null)[];\n}\n"; want != report.Units[0].Text {
		t.Fatalf("text: %q", report.Units[0].Text)
	}
}

func TestLowerBodyPieceWithoutDepth(t *testing.T) {
	profile := bundledProfile("typescript")
	unit := func(piece any) map[string]any {
		return map[string]any{"aontu": map[string]any{"Code": map[string]any{"units": []any{map[string]any{
			"path": "a.ts", "lang": "typescript",
			"decls": []any{map[string]any{
				"k": "func", "name": "f", "params": []any{},
				"body": map[string]any{"k": "frag", "n": []any{
					piece, map[string]any{"k": "blank"}, "done()",
				}},
			}},
		}}}}}
	}
	opts := &RenderOptions{Profiles: []map[string]any{profile}}
	bare := RenderValue(unit(map[string]any{"k": "line", "n": []any{"go()"}}), opts)
	at0 := RenderValue(unit(map[string]any{"k": "line", "at": int64(0), "n": []any{"go()"}}), opts)
	if bare.Verdict != at0.Verdict || 1 != len(bare.Units) || 1 != len(at0.Units) {
		t.Fatalf("bare: %+v\nat0: %+v", bare, at0)
	}
	if bare.Units[0].Text != at0.Units[0].Text {
		t.Fatalf("bare: %q\nat0: %q", bare.Units[0].Text, at0.Units[0].Text)
	}
	if !strings.Contains(bare.Units[0].Text, "\n  go()\n\n  done()\n") {
		t.Fatalf("text: %q", bare.Units[0].Text)
	}
}

File: ts/test/lower.test.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


import { test, describe } from 'node:test'
import Assert from 'node:assert'

import { Aontu } from '../dist/aontu'
import { renderValue } from '../dist/render'
import {
  splitWords, caseName, quote, literal, typeExpr,
} from '../dist/lower'


function ctx(family: string, profile: any): any {
  return { profile: { lang: family, ...profile }, family, unit: 'u', lossy: [] }
}


describe('lower', () => {

  test('words-split-at-the-ascii-boundaries-and-non-ascii-rides', () => {
    Assert.deepStrictEqual(splitWords('HTTPServer2Go'), ['HTTP', 'Server', '2', 'Go'])
    Assert.deepStrictEqual(splitWords('utf8_string-value'), ['utf', '8', 'string', 'value'])
    Assert.deepStrictEqual(splitWords('naïveName'), ['naïve', 'Name'])
    Assert.deepStrictEqual(splitWords('a b'), ['a', 'b'])
    Assert.deepStrictEqual(splitWords('__'), [])
    // A code point at or above U+0080 is never split and never
    // converted: it rides into its word, whatever the style.
    Assert.strictEqual(caseName('crème_brûlée', 'pascal', []), 'CrèmeBrûlée')
    Assert.strictEqual(caseName('über2Go', 'snake', []), 'über_2_go')
    Assert.strictEqual(caseName('a-b c', 'kebab', []), 'a-b-c')
    Assert.strictEqual(caseName('a-b c', 'screaming', []), 'A_B_C')
    Assert.strictEqual(caseName('ledgerId', 'camel', ['ID']), 'ledgerID')
    Assert.strictEqual(caseName('id', 'camel', ['ID']), 'id')
    Assert.strictEqual(caseName('__', 'pascal', []), '__')
    Assert.strictEqual(caseName('AsIs_x', 'as-is', []), 'AsIs_x')
  })

  test('quote-follows-the-profile-quote-and-table', () => {
    const profile = { str: { quote: "'", escape: { '10': '\\n' } } }
    Assert.strictEqual(quote('it\'s "x"\n', profile),
      "'it\\u0027s \"x\"\\n\\u0002'")
    // No str at all: the double quote, and a bare table.
    Assert.strictEqual(quote('a"b\\c', {}), '"a\\u0022b\\u005cc"')
  })

  test('literals-spell-as-the-family-does', () => {
    const go = ctx('go', {})
    const ts = ctx('typescript', {})
    Assert.strictEqual(literal(1.5, go), '1.5')
    Assert.strictEqual(literal(null, go), 'nil')
    Assert.strictEqual(literal(null, ts), 'null')
    Assert.strictEqual(literal(false, ts), 'false')
    Assert.strictEqual(literal('s', ts), '"s"')
  })

  test('a-literal-set-in-go-takes-the-shared-primitive', () => {
    const go = ctx('go', {})
    const prim = (of: any[]) => typeExpr({ k: 'lit', of }, go, '$').text
    Assert.strictEqual(prim([1.5, 2]), 'float')
    Assert.strictEqual(prim([1, 2]), 'int')
    Assert.strictEqual(prim([true]), 'bool')
    Assert.strictEqual(prim([null]), 'null')
    Assert.strictEqual(prim(['a', 1]), 'any')
    Assert.strictEqual(go.lossy.length, 5)
  })

  test('a-profile-without-type-forms-falls-back-to-the-names', () => {
    // A supplied profile may name a lowering and no forms: the
    // primitive is its own name, and a form is open-less and close-less.
    const ts = ctx('typescript', {})
    Assert.strictEqual(typeExpr({ k: 'prim', prim: 'int' }, ts, '$').text, 'int')
    Assert.strictEqual(typeExpr(
      { k: 'list', of: { k: 'prim', prim: 'int' } }, ts, '$').text, 'int')
    Assert.strictEqual(typeExpr(
      { k: 'map', key: { k: 'prim', prim: 'string' }, of: { k: 'prim', prim: 'int' } },
      ts, '$').text, 'string, int')
    Assert.strictEqual(typeExpr(
      { k: 'union', of: [{ k: 'prim', prim: 'a' }, { k: 'prim', prim: 'b' }] },
      ts, '$').text, 'a | b')
  })

  test('the-paren-rule-puts-a-lower-precedence-inner-in-parens', () => {
    // The vocabulary keeps a container to leaves, so a list of a
    // nullable reaches the fold only through renderValue -- where the
    // TypeScript forms say (string | null)[] and not string | null[].
    const profile = new Aontu().generate('@"aontu:render/lang/typescript"').aontu.profile
    const report = renderValue({
      aontu: { Code: {
        units: [{
          path: 'a.ts', lang: 'typescript',
          decls: [{
            k: 'record', name: 'T', open: false, check: [], fields: [{
              name: 'a', optional: false,
              type: { k: 'list', of: { k: 'opt', of: { k: 'prim', prim: 'string' } } },
            }],
          }],
        }],
      },
    } }, { profiles: [profile] })
    Assert.strictEqual(report.verdict, 'ok')
    Assert.strictEqual(report.units[0].text,
      'export interface T {\n  a: (string | null)[];\n}\n')
  })

  test('a-body-piece-without-a-depth-nests-as-depth-zero-does', () => {
    const profile = new Aontu().generate('@"aontu:render/lang/typescript"').aontu.profile
    const unit = (piece: any) => ({
      aontu: { Code: {
        units: [{
          path: 'a.ts', lang: 'typescript',
          decls: [{
            k: 'func', name: 'f', params: [],
            body: { k: 'frag', of: [piece, { k: 'blank' }, 'done()'] },
          }],
        }],
      } },
    })
    const bare = renderValue(unit({ k: 'line', of: ['go()'] }), { profiles: [profile] })
    const at0 = renderValue(unit({ k: 'line', at: 0, of: ['go()'] }), { profiles: [profile] })
    Assert.strictEqual(bare.verdict, at0.verdict)
    Assert.strictEqual(bare.units.length, 1)
    Assert.strictEqual(bare.units[0].text, at0.units[0].text)
    Assert.ok(bare.units[0].text.includes('\n  go()\n\n  done()\n'))
  })
})
